src/libmongoc/tests/unified/runner.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "json-test.h"
+#include "TestSuite.h"
+#include "test-conveniences.h"
+#include "test-libmongoc.h"
+#include <signal.h>
+
+/* The test context is a global storing state for each test.
+ * When an assertion fails, causing an abort signal, the test
+ * context can be logged. */
+struct {
+   bson_t *test;
+   const char *description;
+   bson_t command_started_events;
+   bson_t command_succeeded_events;
+   bson_t command_failed_events;
+} test_ctx;
+
+static void
+test_ctx_init (bson_t *test)
+{
+   test_ctx.test = test;
+   test_ctx.description = bson_lookup_utf8 (test, "description");
+   bson_init (&test_ctx.command_started_events);
+   bson_init (&test_ctx.command_succeeded_events);
+   bson_init (&test_ctx.command_failed_events);
+}
+
+static void
+test_ctx_cleanup ()
+{
+   bson_destroy (&test_ctx.command_started_events);
+   bson_destroy (&test_ctx.command_succeeded_events);
+   bson_destroy (&test_ctx.command_failed_events);
+}
+
+static void
+handle_abort (int signo)
```

Comment:
I want to solve the problem of getting an assertion error in a spec test, and not knowing which operation failed (like an error "assert failed: 3 != 2"). Instead of changing every assertion, this captures the abort signal and prints out state from the global test context.

Reply:
This is definitely a lot more helpful.